### PR TITLE
Page rank algorithm support

### PR DESCRIPTION
**Pull request by geosarr**

### CODE REVIEW

File: src/algo/mod.rs [hunk 1]
```diff
@@ -901,3 +903,42 @@ macro_rules! impl_bounded_measure_float(
 );
 
 impl_bounded_measure_float!(f32, f64);
+
+pub trait UnitMeasure:
```
Comment (ABorgna) [inline src/algo/mod.rs:907]: Add docs to the trait

File: src/algo/page_rank.rs [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+/// let expected_ranks = vec![0.14685437, 0.20267677, 0.22389607, 0.27971846, 0.14685437];
+/// assert_eq!(expected_ranks, output_ranks);
+/// ```
+pub fn page_rank<G, D>(graph: G, damping_factor: D, nb_iter: usize) -> Vec<D>
+where
+    G: NodeCount + IntoEdges + NodeIndexable,
+    D: UnitMeasure + Copy,
+{
+    let node_count = graph.node_count();
+    assert!(node_count > 0, "Graph must have nodes.");
```
Comment (ABorgna) [inline src/algo/page_rank.rs:59]: We could return an empty vector here instead of panicking.

File: src/algo/page_rank.rs [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+        out_degree = out_degree + D::one();
+        if edge.target() == node_v {
+            flag_points_to = true;
+        }
+        out_edge = out_edges.next();
+    }
+    (out_degree, flag_points_to)
+}
+/// \[Generic\] Parrallel Page Rank algorithm.
+/// ```
```
Comment (ABorgna) [inline src/algo/page_rank.rs:119]: Broken docs.

**Review by ABorgna — CHANGES_REQUESTED** [review]
Thanks!
I just have some change request.

File: src/algo/page_rank.rs [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+    damping_factor: D,
+    nb_iter: usize,
+    tol: Option<D>,
+) -> Vec<D>
+where
+    G: NodeCount + IntoEdges + NodeIndexable + std::marker::Sync,
+    D: UnitMeasure + Copy + std::marker::Send + std::marker::Sync,
+{
+    let node_count = graph.node_count();
+    assert!(node_count > 0, "Graph must have nodes.");
```
Comment (ABorgna) [inline src/algo/page_rank.rs:133]:
```suggestion
    if node_count == 0 {
        return vec![];
    }
```

File: src/algo/page_rank.rs [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+    while let Some(edge) = out_edge {
+        out_degree = out_degree + D::one();
+        if edge.target() == node_v {
+            flag_points_to = true;
+        }
+        out_edge = out_edges.next();
+    }
+    (out_degree, flag_points_to)
+}
+/// \[Generic\] Parrallel Page Rank algorithm.
```
Comment (ABorgna) [inline src/algo/page_rank.rs:120]:
```suggestion
/// \[Generic\] Parallel Page Rank algorithm.
///
/// See [`page_rank`].
```

**Review by ABorgna — APPROVED** [review]

**Comment by geosarr** [comment]
Thanks @ABorgna

**Pull request merged**